File: src/DataScience/R/patterns/mediator.R
# Mediator: colleagues communicate through a coordinator, not directly.
events <- character()

routes <- new.env(parent = emptyenv())
routes$inventory <- function(sender, message) {
  events <<- c(events, paste0("inventory<-", sender, ":", message))
}
routes$payment <- function(sender, message) {
  events <<- c(events, paste0("payment<-", sender, ":", message))
}

mediator_send <- function(sender, recipient, message) {
  if (!exists(recipient, envir = routes, inherits = FALSE)) {
    stop(paste("unknown colleague", recipient))
  }
  get(recipient, envir = routes, inherits = FALSE)(sender, message)
}

payment <- function(message) mediator_send("payment", "inventory", message)
inventory <- function(message) mediator_send("inventory", "payment", message)

payment("paid")
inventory("reserved")
stopifnot(identical(events, c("inventory<-payment:paid", "payment<-inventory:reserved")))

rejected <- inherits(
  try(mediator_send("payment", "unknown", "ignored"), silent = TRUE),
  "try-error"
)
stopifnot(rejected)

File: src/DataScience/Julia/patterns/mediator.jl
# Mediator — colleagues communicate through a routing authority rather than directly.
mutable struct CheckoutMediator
    colleagues::Dict{String, Function}
end

CheckoutMediator() = CheckoutMediator(Dict{String, Function}())

function register!(mediator::CheckoutMediator, name::String, receiver::Function)
    mediator.colleagues[name] = receiver
    mediator
end

function send(mediator::CheckoutMediator, sender::String, recipient::String, message::String)
    receiver = get(mediator.colleagues, recipient, nothing)
    receiver === nothing && return (false, "unknown colleague: $recipient")
    (true, receiver(sender, message))
end

payment(sender::String, message::String) = "payment<-$sender:$message"
inventory(sender::String, message::String) = "inventory<-$sender:$message"

function verify_mediator()
    mediator = CheckoutMediator()
    register!(mediator, "payment", payment)
    register!(mediator, "inventory", inventory)

    paid = send(mediator, "payment", "inventory", "paid")
    reserved = send(mediator, "inventory", "payment", "reserved")
    missing = send(mediator, "payment", "shipping", "dispatch")

    @assert paid == (true, "inventory<-payment:paid")
    @assert reserved == (true, "payment<-inventory:reserved")
    @assert missing == (false, "unknown colleague: shipping")
    true
end

verify_mediator()
println("Julia Mediator: passed")
